File: kmp.cc
#include<string>
#include<iostream>

using namespace std::cout;

void ComputeNextTable(string &needle, int next[]) {
    int pos = 2;
    int pre_pos = 0;
    next[0] = -1;
    next[1] = 0;
    while(pos < needle.length()) {
        if(W[pos-1] == W[pre]) {
            pre ++;
            next[pos] = pre;
            pos++;
        } else {
            if(pre>0) pre = next[pre];
            else {
                next[pos_h] = 0;
                pos_h ++;
            }
        }
    }
}

int KMPAlgorithm(string haystack, string needle) {
    int pos_h = 0;
    int pos_n = 0;
    int next[needle.length()];

    ComputeNextTable(match, next);
    while (pos_h + pos_n < haystack.length()) {
        if(haystack[pos_h + pos_n] == needle[pos_i]) {
            if(pos_n == needle.length() - 1) return pos_h;
            pos_n ++;
        }
        else {
            if(next[pos_n] > -1) {
                pos_h = pos_h + pos_n - next[pos_n];
                pos_n = next[pos_n];
            }
            else {
                pos_n = 0;
                pos_h ++;
            }
        }
    }
    return haystack.length();
}

int main() {
    cout << KMPAlgorithm("abcabcababc", "abababc") << endl;
}

File: Java/KMP.java
class KMP {
  public int match(String needle, String haystack) {
    int[] next = new int[needle.length()];
    calc(needle, next);
    int pn = 0;
    int ph = 0;
    while (ph < haystack.length()) {
      if (haystack.charAt(ph + pn) == needle.charAt(pn)) {
        pn ++;
        if (pn == needle.length()) return ph;
      } else {
        if (pn > 0) {
          ph = ph + pn - next[pn];
          pn = next[pn];
        } else {
          ph ++;
        }
      }
    }
    return -1;
  }

  private void calc(String needle, int [] next) {
    next[0] = -1;
    next[1] = 0;
    int p = 2;
    int q = 0;
    while (p < needle.length()) {
      if (needle.charAt(p - 1) == needle.charAt(q)) {
        next[p] = q + 1;
        q ++;
        p ++;
      } else {
        if (next[q] >= 0) {
          q = next[q];
        } else {
          next[p] = 0;
          p = p + 1;
        }
      }
    }
  }
}


// a b c d e f
//
